reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamUnicastPublisher.scala

```diff
@@ -25,35 +25,43 @@ package reactivestreams
 
 import cats.effect.kernel._
 import cats.effect.std.Dispatcher
-import cats.syntax.all._
 
 import org.reactivestreams._
 
+import scala.util.control.NoStackTrace
+
 /** Implementation of a `org.reactivestreams.Publisher`
   *
   * This is used to publish elements from a `fs2.Stream` to a downstream reactivestreams system.
   *
+  * @note Not longer unicast, this Publisher can be reused for multiple Subscribers.
+  *
   * @see [[https://github.com/reactive-streams/reactive-streams-jvm#1-publisher-code]]
   */
 final class StreamUnicastPublisher[F[_]: Async, A] private (
     val stream: Stream[F, A],
-    startDispatcher: Dispatcher[F],
-    requestDispatcher: Dispatcher[F]
+    startDispatcher: Dispatcher[F]
 ) extends Publisher[A] {
 
-  @deprecated("Use StreamUnicastPublisher.apply", "3.4.0")
-  def this(stream: Stream[F, A], dispatcher: Dispatcher[F]) = this(stream, dispatcher, dispatcher)
+  /** Constructor is defined twice
+    *  @deprecated("Use StreamUnicastPublisher.apply", "3.4.0")
+    *  def this(stream: Stream[F, A], dispatcher: Dispatcher[F]) =
+    *    this(stream, dispatcher)
+    */
 
   def subscribe(subscriber: Subscriber[_ >: A]): Unit = {
     nonNull(subscriber)
-    startDispatcher.unsafeRunAndForget {
-      StreamSubscription(subscriber, stream, startDispatcher, requestDispatcher)
-        .flatMap { subscription =>
-          Sync[F].delay {
-            subscriber.onSubscribe(subscription)
-            subscription.unsafeStart()
-          }
-        }
+    try
+      startDispatcher.unsafeRunAndForget(
+        StreamSubscription.subscribe(stream, subscriber)
+      )
+    catch {
+      case _: IllegalStateException =>
+        subscriber.onSubscribe(new Subscription {
+          override def cancel(): Unit = ()
+          override def request(x$1: Long): Unit = ()
+        })
+        subscriber.onError(StreamUnicastPublisher.CanceledStreamPublisherException)
```

Comment:
Why is this calling `.onSubscribe` first? The Javadoc for `Publisher` says this:

> If already subscribed, or the attempt to subscribe fails due to policy violations or errors, the `Subscriber`'s `onError` method is invoked with an `IllegalStateException`. Otherwise, the `Subscriber`'s `onSubscribe` method is invoked

https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.Publisher.html#subscribe-java.util.concurrent.Flow.Subscriber-

So it seems unnecessary.

Reply:
@armanbilge specification `9` for `Publisher`: https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md

> Publisher.subscribe MUST call onSubscribe on the provided Subscriber prior to any other signals to that Subscriber and MUST [return normally](https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md#term_return_normally), except when the provided Subscriber is null in which case it MUST throw a java.lang.NullPointerException to the caller, for all other situations the only legal way to signal failure (or reject the Subscriber) is by calling onError (after calling onSubscribe).

Actually, I added this because a test failed because of that.